### PR TITLE
Fix/14684: IOU details Last transaction appears at the bottom in the list until logout and login back

### PR DESCRIPTION
**Pull request by tienifr**
### Details

we aren't sorting by timestamp (the created field in reportActions) and we also aren't filtering out invalid report actions (that have sequenceNumber field) to prevent bugs during the migration from sequenceNumber -> reportActionID. In the ReportActionView Component, we just had a function to handle that so we will move this function into ReportActionUtils and use it in both IOUTransaction and ReportActionView
### Fixed Issues

$ https://github.com/Expensify/App/issues/14684
$ https://github.com/Expensify/App/issues/14684#issuecomment-1424554836

### Tests

1. Open the app and log in with user A
2. Open the app and log in with user B on another device
3. On user A request money from user B two times any amount
4. On user B check the IOU action box and make sure that the latest amount request is showing
5. Open the IOU detail by pressing the IOU action box
6. Verify the last transaction amount requested is on the top of the list of transaction
- [x] Verify that no errors appear in the JS console

### QA Steps

1. Open the app and log in with user A
2. Open the app and log in with user B on another device
3. On user A request money from user B two times any amount
4. On user B check the IOU action box and make sure that the latest amount request is showing
5. Open the IOU detail by pressing the IOU action box
6. Verify the last transaction amount requested is on the top of the list of transaction
- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/113963320/218916878-4099b289-dafe-4212-bae1-10351729b1f5.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/113963320/218916853-e669720b-f7aa-4684-aeeb-b72b8278a49f.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/113963320/218917708-641ab56d-61a7-4d5f-a5ee-6f37aeff8ae8.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/113963320/218916932-d25eb944-037a-4a8c-93cc-8fe84089711a.mp4

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/113963320/218917499-74f30e5a-6b6c-4786-96d1-1cc77bba910c.mp4

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/113963320/218918823-f3d97e4d-b9e8-4c8b-a2f0-b1d950c108c8.mp4

</details>

### CODE REVIEW

File: src/pages/home/report/ReportActionsView.js
Comment (mollfpr) [inline src/pages/home/report/ReportActionsView.js:268]: @tienifr Any reason why we delete the log? It's just got updated 2 days ago in this PR https://github.com/Expensify/App/pull/15146

Comment (tienifr) [reply src/pages/home/report/ReportActionsView.js:268]: @mollfpr Because this log is a new update, I don't see that before. I will add it again

File: src/libs/ReportActionsUtils.js
Comment (mollfpr) [inline src/libs/ReportActionsUtils.js:204]: @tienifr Can we just return `filterReportActionsForDisplay(sortedReportActions)` rather than assign it to variable again?

Comment (tienifr) [reply src/libs/ReportActionsUtils.js]: @mollfpr I just updated. Thanks for pointing out it

**Comment by mollfpr** [comment]
@tienifr I think the test case could be clearer. How about:

1. Open the app and log in with user A
2. Open the app and log in with user B
3. On user A request money from user B two times any amount
4. On user B check the IOU action box and make sure that the latest amount request is showing
5. Open the IOU detail by pressing the IOU action box
6. Verify the last transaction amount requested is on the top of the list transaction

**Comment by tienifr** [comment]
@mollfpr Thanks for your comment. I updated

File: src/libs/ReportActionsUtils.js
Comment (mollfpr) [inline src/libs/ReportActionsUtils.js:183]: `reportActions` is an object, not an Array

```suggestion
 * @param {Object} reportActions
```

File: src/libs/ReportActionsUtils.js
Comment (mollfpr) [inline src/libs/ReportActionsUtils.js:186]: The comment for the function needs to explain why we need this function.

**Comment by mollfpr** [comment]
@tgolen Should we add the test for `getSortedReportActionsForDisplay`? We have the `ReportActionsUtilsTest` file but only two functions is added for the test.

File: src/libs/ReportActionsUtils.js
Comment (mollfpr) [inline src/libs/ReportActionsUtils.js:181]: @tgolen Any suggestion about this comment? I personally think this is enough to explain what the function is, but probably you have others thoughts 😅

**Comment by mollfpr** [comment]
@tienifr Could you resolve the conflicts? Thanks!

**Comment by tienifr** [comment]
Hi, @mollfpr  I just resolved the conflict, but there are some failure tests in the main branch so after merging the main branch this PR also gets failure tests. I will merge the main branch again after the failure test is fixed. @tgolen Could you help to take a look at this PR?

**Comment by tienifr** [comment]
Hi @mollfpr @tgolen, I just merged the main branch again, and all successful. Please help to check the PR. Thanks

**Comment by mollfpr** [comment]
Sorry for the delay, and thank you for updating the PR! I’ll review it within an hour.

**Review by mollfpr — DISMISSED** [review]
LGTM 👍

All yours @tgolen

**Comment by tienifr** [comment]
We got @mollfpr's approval. Waiting for your review @tgolen. Thank you!

File: src/libs/ReportActionsUtils.js
Comment (tgolen) [inline src/libs/ReportActionsUtils.js:213]: I don't think this is referenced anywhere outside this file now, so it's export can be removed.

File: src/libs/ReportActionsUtils.js
Comment (tgolen) [inline src/libs/ReportActionsUtils.js:208]: Since this is the only reference to `filterReportActionsForDisplay`, I guess it doesn't need it's own method anymore and the logic can be moved inside of this method. Would you agree?

File: src/libs/ReportActionsUtils.js
Comment (tgolen) [reply src/libs/ReportActionsUtils.js:186]: Sure, I have a few thoughts!

1. The hack doesn't need to be mentioned (it's mentioned in its own comment below)
2. The part about being a combination of two other methods doesn't seem necessary since it's obvious from looking at the code
3. I think the comment should describe _WHY_ the method needs to be a combination of two other methods (but I have a suggestion below which will remove one of the methods)

I suggest:
> This method returns the report actions that are ready for display in the ReportActionsView. The report actions need to be sorted by created timestamp first, and reportActionID second to ensure they will always be displayed in the same order (in case multiple actions have the same timestamp). This is all handled with getSortedReportActions() which is used by several other methods to keep the code DRY.

File: src/pages/home/report/ReportActionsView.js
Comment (tgolen) [inline src/pages/home/report/ReportActionsView.js:135]: Not-a-blocker: I really dislike seeing these side-effects in this method. It would be really nice to see if there is a way to move these somewhere else at some point. No need to do it in this PR, but if you have some ideas, it might be nice to open a GH to do this at some point.

**Review by tgolen — CHANGES_REQUESTED** [review]

File: src/libs/ReportActionsUtils.js
Comment (tienifr) [reply src/libs/ReportActionsUtils.js:213]: I can see that both `getSortedReportActions` and `filterReportActionsForDisplay` are being used outside this file.
For `getSortedReportActions`: https://github.com/Expensify/App/blob/6f2705807eadea43972bc3fdef69237c8c24277a/src/libs/actions/User.js#L270
For `filterReportActionsForDisplay` https://github.com/Expensify/App/blob/6f2705807eadea43972bc3fdef69237c8c24277a/tests/unit/ReportActionsUtilsTest.js#L141

File: src/libs/ReportActionsUtils.js
Comment (tienifr) [reply src/libs/ReportActionsUtils.js:208]: `filterReportActionsForDisplay` is being used in https://github.com/Expensify/App/blob/6f2705807eadea43972bc3fdef69237c8c24277a/tests/unit/ReportActionsUtilsTest.js#L141 so I think we shouldn't remove it's own method.

File: src/pages/home/report/ReportActionsView.js
Comment (tienifr) [reply src/pages/home/report/ReportActionsView.js:135]: Thanks for the suggestion, I'll think about it separately from this PR.

File: src/libs/ReportActionsUtils.js
Comment (tienifr) [reply src/libs/ReportActionsUtils.js:186]: Thank you for your suggestion!
Duo to what I have mentioned in other replies, we should not remove `filterReportActionsForDisplay` method since it's being used in other places. I will update the comment as you suggested
> 3. I think the comment should describe WHY the method needs to be a combination of two other methods (but I have a suggestion below which will remove one of the methods)

File: src/libs/ReportActionsUtils.js
Comment (tienifr) [reply src/libs/ReportActionsUtils.js:214]: Hi @tgolen Could you help to check my comment? Thanks

File: src/libs/ReportActionsUtils.js
Comment (mollfpr) [reply src/libs/ReportActionsUtils.js:209]: That's a great idea from @tgolen! I'm inclined to have `filterReportActionsForDisplay` removed from export and the unit test. Then we can add the `getSortedReportActionsForDisplay` for the unit test. @tienifr Are you okay with that?

Comment (tienifr) [reply src/libs/ReportActionsUtils.js]: @mollfpr @tgolen I just removed the UTs of filterReportActionsForDisplay function and add UTs for getSortedReportActionsForDisplay function with the same case in UTs of filterReportActionsForDisplay function because we had UTs for getSortedReportActions function before and this hack logic in the below image will be removed in future. So we just need to add the UTs like in the previous filterReportActionsForDisplay function

Screenshot 2023-02-22 at 23 00 24

**Review by tgolen — APPROVED** [review]
This looks really great now, thank you for making those additional changes!

**Pull request merged**
